From c3595c2e62300c04b72e28333dc66da2bd0c695c Mon Sep 17 00:00:00 2001
From: Li Jun <B47624@freescale.com>
Date: Sat, 27 Sep 2014 20:19:30 +0800
Subject: [PATCH 0121/5242] MLK-9637-4 usb: chipidea: otg: test: add
 otg_srp_reqd timer for start srp

commit  e4361acd39d7e07c18bcc7ff2ed78ea0bbe82794 from
https://source.codeaurora.org/external/imx/linux-imx.git

This patch adds a timer for chipidea otg fsm driver to start data pulse
after A host ends the session.

Acked-by: Peter Chen <peter.chen@freescale.com>
Signed-off-by: Li Jun <b47624@freescale.com>
(cherry picked from commit 35e92b59080b58925c3b96e36511056cb51d4a00)
(cherry picked from commit eeded24ce2d20b1c19d5c666e916f41995cf2385)
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/usb/chipidea/otg_fsm.c |   19 +++++++++++++++++++
 drivers/usb/chipidea/otg_fsm.h |    3 +++
 include/linux/usb/otg-fsm.h    |    1 +
 3 files changed, 23 insertions(+)

diff --git a/drivers/usb/chipidea/otg_fsm.c b/drivers/usb/chipidea/otg_fsm.c
index 03081d4..0251362 100644
--- a/drivers/usb/chipidea/otg_fsm.c
+++ b/drivers/usb/chipidea/otg_fsm.c
@@ -216,6 +216,7 @@
 	TB_SSEND_SRP,
 	TA_DP_END,
 	TA_TST_MAINT,
+	TB_SRP_REQD,
 	0,
 };
 
@@ -385,6 +386,21 @@ static int a_tst_maint_tmout(struct ci_hdrc *ci)
 }
 
 /*
+ * otg_srp_reqd feature
+ * After A(PET) turn off vbus, B(UUT) should start this timer to do SRP
+ * when the timer expires.
+ */
+static int b_srp_reqd_tmout(struct ci_hdrc *ci)
+{
+	ci->fsm.otg_srp_reqd = 0;
+	if (ci->fsm.otg->state == OTG_STATE_B_IDLE) {
+		ci->fsm.b_bus_req = 1;
+		return 0;
+	}
+	return 1;
+}
+
+/*
  * Keep this list in the same order as timers indexed
  * by enum otg_fsm_timer in include/linux/usb/otg-fsm.h
  */
@@ -403,6 +419,7 @@ static int (*otg_timer_handlers[])(struct ci_hdrc *) = {
 	b_ssend_srp_tmout,	/* B_SSEND_SRP */
 	a_dp_end_tmout,		/* A_DP_END */
 	a_tst_maint_tmout,	/* A_TST_MAINT */
+	b_srp_reqd_tmout,	/* B_SRP_REQD */
 	NULL,			/* HNP_POLLING */
 };
 
@@ -809,6 +826,8 @@ irqreturn_t ci_otg_fsm_irq(struct ci_hdrc *ci)
 					ci_otg_add_timer(ci, B_SSEND_SRP);
 				if (fsm->b_bus_req)
 					fsm->b_bus_req = 0;
+				if (fsm->otg_srp_reqd)
+					ci_otg_add_timer(ci, B_SRP_REQD);
 			} else {
 				ci->vbus_glitch_check_event = true;
 			}
diff --git a/drivers/usb/chipidea/otg_fsm.h b/drivers/usb/chipidea/otg_fsm.h
index 54ba8db..867c40c 100644
--- a/drivers/usb/chipidea/otg_fsm.h
+++ b/drivers/usb/chipidea/otg_fsm.h
@@ -65,6 +65,9 @@
 #define TB_SSEND_SRP         (1500)	/* minimum 1.5 sec, section:5.1.2 */
 
 #define TB_AIDL_BDIS         (20)	/* 4ms ~ 150ms, section 5.2.1 */
+#define TB_SRP_REQD          (2000)	/* For otg_srp_reqd to start data
+					 * pulse after A(PET) turn off v-bus
+					 */
 
 #if IS_ENABLED(CONFIG_USB_OTG_FSM)
 
diff --git a/include/linux/usb/otg-fsm.h b/include/linux/usb/otg-fsm.h
index 1655fdc..718d93a 100644
--- a/include/linux/usb/otg-fsm.h
+++ b/include/linux/usb/otg-fsm.h
@@ -56,6 +56,7 @@ enum otg_fsm_timer {
 	B_SSEND_SRP,
 	A_DP_END,
 	A_TST_MAINT,
+	B_SRP_REQD,
 
 	NUM_OTG_FSM_TIMERS,
 };
-- 
1.7.9.5

